### DISCUSSION TITLE
How to add footnotes

### DISCUSSION THREAD

**Discussion by fbanning** [typst/typst#222] [Q&A] [answered]
How to add footnotes to a document? I couldn't find anything in the typst docs about this and would have bet my only pair of shoelaces on this already existing as a predefined function in typst.

**Comment by laurmaedje** [comment, marked as answer]
This is on the roadmap. We know that it is very important, we just didn't come to it yet.

**Reply by IcedQuinn** [reply]
Sidenotes would be nice too, for the Tufte enjoyers among us :)

**Reply by fbanning** [reply]
Oh, that's great. Thanks for the info.

> Sidenotes would be nice too, for the Tufte enjoyers among us :)

💯 !

**Reply by saadulkh** [reply]
So, till it gets implemented in Typst. I've created a little ![typst-notes](https://github.com/saadulkh/typst-notes) library for it.

**Reply by rwmpelstilzchen** [reply]
For completeness’ sake, and for anyone who reads this discussion later in time, there are at least five additional packages/templates that provide sidenotes by now (2026-04):
https://typst.app/universe/package/marge
https://typst.app/universe/package/marginalia
https://typst.app/universe/package/tufte-memo
https://typst.app/universe/package/tufted
https://typst.app/universe/package/toffee-tufte

**Comment by sirodoht** [comment]
I want to typeset a book using typst and I want to have both footnotes and page numbers. Can I use @saadulkh's solution to do this? Can I somehow change [this code](https://github.com/saadulkh/typst-notes/blob/main/notes.typ) to include both page numbers and notes?

**Reply by vangola-silver** [reply]
You could modify the display function like this to also show the page number. However, the page number will not be a constant place in the page but move with each added footnote downwards. If you have more than 4 footnotes in one page, the page number will disappear. 
I am afraid we have no other way but to wait until footnote will be added officially or the authers will share their secret on how they introduced the footnotes in their master's theses.
`
#let display() = {
  set text(size: 8pt)
  locate(loc => {
    let pg = loc.page()
    let dict = notes.final(loc)
    let vec = ()
    if str(pg) in dict.keys() {
      vec = dict.at(str(pg))
    }
       
    
    if vec.len() > 0 {
            line(length: 100%)
            for i in range(0, vec.len()) {
                super[#{i + 1}] + h(1pt) + vec.at(i)
                linebreak()
                
            }
    }

        set text(size: 20pt)
        super[#align(center, [#counter(page).display()])]

  })
}
`

**Reply by saadulkh** [reply]
> I want to have both footnotes and page numbers.

You can modify your `set page`'s footer parameter like this: (see [example.typ](https://github.com/saadulkh/typst-notes/blob/main/example.typ))
```
        footer: {
            grid(
                columns: (1fr, auto),
                notes.display(),
                {
                    v(8pt)
                    counter(page).display()
                }
            )
        }, // Footnotes
```

Tho, keep in mind that it's a solution for the time being and we are all waiting for the official implementation.
